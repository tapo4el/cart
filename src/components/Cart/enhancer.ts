import { connect } from 'react-redux';
import { Dispatch } from 'redux';

import { AppState, Bike, AmountPayload } from '../../types';
import {
    getCartItems, getTotalPrice,
} from '../../selectors';
import { removeItem, clearCart, changeAmount } from '../../actions';

const mapStateToProps = (state: AppState): StateProps => ({
    items: getCartItems(state),
    totalPrice: getTotalPrice(state),
});

const mapDispatchToProps = (dispatch: Dispatch) => ({
    removeItem: (payload: string) => dispatch(removeItem(payload)),
    clearCart: () => dispatch(clearCart()),
    changeAmount: (payload: AmountPayload) => dispatch(changeAmount(payload)),
});

export default connect(mapStateToProps, mapDispatchToProps);

interface StateProps { items: Bike[]; totalPrice: number }
export type CartProps = StateProps & ReturnType<typeof mapDispatchToProps>
