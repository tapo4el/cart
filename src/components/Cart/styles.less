@import "../../vars.less";

.cart {
  max-width: 960px;
  margin: 20px auto;
  font-family: @default-font-family;
  font-size: @default-font-size;
  border: @default-border;
  padding: @default-padding*2;
  border-radius: 10px;

  &>.header {
    font-size: @default-font-size*2;
    padding-bottom: @default-padding*2;
  }

  &>.footer {
    border-top: @default-border;
    padding-top: @default-padding*2;
    text-align: right;
  }

  .totalPrice {
    font-size: @default-header-size;
    padding: @default-padding 0;
    text-align: right;
  }
}

.item {
  display: flex;
  padding: @default-padding;
  border-top: @default-border;

  .image {
    flex: 0 0 auto;
  }
  .content {
    flex: 1 1 auto;
    padding: @default-padding;

    &>div {
      padding-bottom: 5px;
    }

    &>.header {
      font-size: @default-header-size;
    }

    &>.footer {
      text-align: right;
    }

    .amount {
      button:first-child {
        margin-left: 10px;
      }
      button {
        display: inline-block;
        border: @default-border;
        width: 24px;
        text-align: center;
        cursor: pointer;
        border-radius: 5px;
        outline: none;
      }
      i {
        display: inline-block;
        padding: 0 8px;
      }
    }
  }
}

.removeButton {
  background: #db2828;
  cursor: pointer;
  color: #ffffff;
  padding: 10px 20px;
  font-size: 14px;
  font-weight: bold;
  border: none;
  border-radius: 5px;
  outline: none;
}
