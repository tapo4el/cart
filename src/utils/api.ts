import { Bike } from '../types';

export interface BikesAPI {
    (): Promise<Bike[]>;
}

export default async function getBikes(): Promise<Bike[]> {
    const response = await fetch('/data', { method: 'GET' });
    return response.json();
}
