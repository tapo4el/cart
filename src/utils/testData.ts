import {
    Bike,
} from '../types';


export const response: Bike[] = [
    {
        id: '1',
        company: 'Soma',
        model: 'Doublecross',
        price: '100',
        date: 'Feb 24, 2013',
        description: 'Soma Doublecross frameset carbon fork like new',
        picture: 'https://cdn.bike-discount.de/media/org/orgb_C/orgid_26/thumbs/328964_2936140.jpg',
        Link: 'Link To website',
    },
    {
        id: '2',
        company: 'Soma',
        model: 'Smoothie',
        price: '200',
        date: 'Feb 18, 2013',
        description: '2012 Soma Smoothie Frameset, 52 cm',
        picture: 'https://cdn.bike-discount.de/media/org/orgb_R/orgid_33/thumbs/450554_3853977.jpg',
        Link: 'Link To website',
    },
    {
        id: '3',
        company: 'Soma',
        model: 'DELANCEY',
        price: '300',
        date: 'Jan 25, 2013',
        description: 'SOMA DELANCEY Green Fixie Frame and Fork NEW 54cm x 54cm SoundCycles Los Angeles',
        picture: 'https://cdn.bike-discount.de/media/org/orgb_R/orgid_33/thumbs/486175_4045883.jpg',
        Link: 'Link To website',
    },
];

export const cartInitialState: Bike[] = [];
export const cartFinalState = response.map(elem => ({ ...elem, amount: '1' }));

export const initialAppState = {
    cart: cartInitialState,
};
export const finalAppState = {
    cart: cartFinalState,
};

// eslint-disable-next-line @typescript-eslint/explicit-function-return-type
export const getCopy = (obj: any) => JSON.parse(JSON.stringify(obj));
