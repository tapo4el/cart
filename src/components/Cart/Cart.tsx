import * as React from 'react';

import './styles.less';

import { CartProps } from './enhancer';
import { Bike } from '../../types';

class Cart extends React.PureComponent<CartProps, null> {
    renderItem(elem: Bike): React.ReactNode {
        const {
            id, company, model, price, picture, description, amount,
        } = elem;
        const { removeItem, changeAmount } = this.props;

        return (
            <div className="item" key={id}>
                <div className="image">
                    <img src={picture} alt={model} />
                </div>
                <div className="content">
                    <div className="header">{model}</div>
                    <div className="description">{`Description: ${description}`}</div>
                    <div className="company">{`Company: ${company}`}</div>
                    <div className="price">{`Price: ${price} $`}</div>
                    <div className="amount">
                        Amount:
                        <button type="button" onClick={() => changeAmount({ id, amount: '1' })}>+</button>
                        <i>{amount}</i>
                        <button type="button" onClick={() => changeAmount({ id, amount: '-1' })}>-</button>
                    </div>
                    <div className="footer">
                        <button
                            type="button"
                            className="removeButton"
                            onClick={() => removeItem(id)}
                        >
                            Remove
                        </button>
                    </div>
                </div>
            </div>
        );
    }

    render(): React.ReactNode {
        const { items, totalPrice, clearCart } = this.props;

        return (
            <div className="cart">
                <div className="header">Cart</div>
                <div className="totalPrice">{`Total price: ${totalPrice} $`}</div>
                { items.map(elem => this.renderItem(elem))}
                <div className="footer">
                    <button
                        type="button"
                        className="removeButton"
                        onClick={clearCart}
                    >
                        Clear Cart
                    </button>
                </div>
            </div>
        );
    }
}

export default Cart;
