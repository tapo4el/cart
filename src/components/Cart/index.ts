import Cart from './Cart';
import enhancer from './enhancer';

export default enhancer(Cart);
