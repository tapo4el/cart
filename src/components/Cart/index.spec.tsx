import * as React from 'react';
import * as renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux';

import Cart from './index';
import { finalAppState, getCopy } from '../../utils/testData';

const mockStore = configureStore();

describe('Cart', () => {
    it('should be rendered correctly', () => {
        const store = mockStore(finalAppState);
        // @ts-ignore
        const component = renderer.create(<Cart store={store} />);
        expect(component.toJSON()).toMatchSnapshot();
    });

    it('should remove item', async () => {
        const store = mockStore(getCopy(finalAppState));
        const wrapper = mount(
            <Provider store={store}><Cart /></Provider>,
        );
        await wrapper.find('.removeButton').at(0).simulate('click');
        expect(store.getActions()).toMatchSnapshot();
    });

    it('should clear cart', async () => {
        const store = mockStore(getCopy(finalAppState));
        const wrapper = mount(
            <Provider store={store}><Cart /></Provider>,
        );
        await wrapper.find('button[data-test-id="clearCart"]').at(0).simulate('click');
        expect(store.getActions()).toMatchSnapshot();
    });

    it('should increase amount of the item', async () => {
        const store = mockStore(getCopy(finalAppState));
        const wrapper = mount(
            <Provider store={store}><Cart /></Provider>,
        );
        await wrapper.find('button[data-test-id="increaseAmount"]').at(0).simulate('click');
        expect(store.getActions()).toMatchSnapshot();
    });
    it('should decrease amount of the item', async () => {
        const store = mockStore(getCopy(finalAppState));
        const wrapper = mount(
            <Provider store={store}><Cart /></Provider>,
        );
        await wrapper.find('button[data-test-id="decreaseAmount"]').at(0).simulate('click');
        expect(store.getActions()).toMatchSnapshot();
    });
});
