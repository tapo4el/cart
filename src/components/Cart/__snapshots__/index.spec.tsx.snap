// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cart should be rendered correctly 1`] = `
<div
  className="cart"
>
  <div
    className="header"
  >
    Cart
  </div>
  <div
    className="totalPrice"
  >
    Total price: 600 $
  </div>
  <div
    className="item"
  >
    <div
      className="image"
    >
      <img
        alt="Doublecross"
        src="https://cdn.bike-discount.de/media/org/orgb_C/orgid_26/thumbs/328964_2936140.jpg"
      />
    </div>
    <div
      className="content"
    >
      <div
        className="header"
      >
        Doublecross
      </div>
      <div
        className="description"
      >
        Description: Soma Doublecross frameset carbon fork like new
      </div>
      <div
        className="company"
      >
        Company: Soma
      </div>
      <div
        className="price"
      >
        Price: 100 $
      </div>
      <div
        className="amount"
      >
        Amount:
        <button
          data-test-id="increaseAmount"
          onClick={[Function]}
          type="button"
        >
          +
        </button>
        <i>
          1
        </i>
        <button
          data-test-id="decreaseAmount"
          onClick={[Function]}
          type="button"
        >
          -
        </button>
      </div>
      <div
        className="footer"
      >
        <button
          className="removeButton"
          onClick={[Function]}
          type="button"
        >
          Remove
        </button>
      </div>
    </div>
  </div>
  <div
    className="item"
  >
    <div
      className="image"
    >
      <img
        alt="Smoothie"
        src="https://cdn.bike-discount.de/media/org/orgb_R/orgid_33/thumbs/450554_3853977.jpg"
      />
    </div>
    <div
      className="content"
    >
      <div
        className="header"
      >
        Smoothie
      </div>
      <div
        className="description"
      >
        Description: 2012 Soma Smoothie Frameset, 52 cm
      </div>
      <div
        className="company"
      >
        Company: Soma
      </div>
      <div
        className="price"
      >
        Price: 200 $
      </div>
      <div
        className="amount"
      >
        Amount:
        <button
          data-test-id="increaseAmount"
          onClick={[Function]}
          type="button"
        >
          +
        </button>
        <i>
          1
        </i>
        <button
          data-test-id="decreaseAmount"
          onClick={[Function]}
          type="button"
        >
          -
        </button>
      </div>
      <div
        className="footer"
      >
        <button
          className="removeButton"
          onClick={[Function]}
          type="button"
        >
          Remove
        </button>
      </div>
    </div>
  </div>
  <div
    className="item"
  >
    <div
      className="image"
    >
      <img
        alt="DELANCEY"
        src="https://cdn.bike-discount.de/media/org/orgb_R/orgid_33/thumbs/486175_4045883.jpg"
      />
    </div>
    <div
      className="content"
    >
      <div
        className="header"
      >
        DELANCEY
      </div>
      <div
        className="description"
      >
        Description: SOMA DELANCEY Green Fixie Frame and Fork NEW 54cm x 54cm SoundCycles Los Angeles
      </div>
      <div
        className="company"
      >
        Company: Soma
      </div>
      <div
        className="price"
      >
        Price: 300 $
      </div>
      <div
        className="amount"
      >
        Amount:
        <button
          data-test-id="increaseAmount"
          onClick={[Function]}
          type="button"
        >
          +
        </button>
        <i>
          1
        </i>
        <button
          data-test-id="decreaseAmount"
          onClick={[Function]}
          type="button"
        >
          -
        </button>
      </div>
      <div
        className="footer"
      >
        <button
          className="removeButton"
          onClick={[Function]}
          type="button"
        >
          Remove
        </button>
      </div>
    </div>
  </div>
  <div
    className="footer"
  >
    <button
      className="removeButton"
      data-test-id="clearCart"
      onClick={[Function]}
      type="button"
    >
      Clear Cart
    </button>
  </div>
</div>
`;

exports[`Cart should clear cart 1`] = `
Array [
  Object {
    "type": "CLEAR_CART",
  },
]
`;

exports[`Cart should decrease amount of the item 1`] = `
Array [
  Object {
    "payload": Object {
      "amount": "-1",
      "id": "1",
    },
    "type": "CHANGE_AMOUNT",
  },
]
`;

exports[`Cart should increase amount of the item 1`] = `
Array [
  Object {
    "payload": Object {
      "amount": "1",
      "id": "1",
    },
    "type": "CHANGE_AMOUNT",
  },
]
`;

exports[`Cart should remove item 1`] = `
Array [
  Object {
    "payload": "1",
    "type": "REMOVE_ITEM",
  },
]
`;
